refactor(sozo): make event parsing logic modular and reusable

Matth26 opened the pull request:
This PR addresses Issue #882, resolves DOJ-108, focusing on leveraging the `cainome` parser for ABI parsing and processing Emitted Events.

## Current Progress
- Successfully integrated the `cainome` parser for parsing ABI and Emitted Events.
- `CoreBasic` and `Array` data structures are correctly parsed.

@glihm, could I have your review on the work done so far to see if it aligns with your expectations?

glihm commented on bin/sozo/src/ops/events.rs:
```diff
@@ -38,126 +37,255 @@ pub async fn execute(
+    }
+}
+
+fn parse_core_basic(
+    cb: &CoreBasic,
+    value: &FieldElement,
+    is_nested: bool,
+) -> Result<String, String> {
+    match cb.type_name().as_str() {
+        "felt252" => {
+            if is_nested {
+                Ok(format!("\"{:#x}\"", value))
+            } else {
+                match parse_cairo_short_string(value) {
+                    Ok(parsed) => Ok(parsed),
+                    Err(_) => Ok(format!("\"{:#x}\"", value)),
+                }
+            }
+        }
+        "bool" => {
+            if *value == FieldElement::ZERO {
+                Ok("false".to_string())
+            } else {
+                Ok("true".to_string())
+            }
+        }
+        "ClassHash" | "ContractAddress" => Ok(format!("{:#x}", value)),
+        "u8" | "u16" | "u32" | "u64" | "u128" | "usize" | "i8" | "i16" | "i32" | "i64" | "i128" => {
+            Ok(value.to_string())
+        }
+        _ => Err(format!("Unsupported CoreBasic type: {}", cb.type_name())),
+    }
+}
```
Would be amazing to have that into cainome, having a `print_value` method on the `CoreBasic` token type.

To allow some configuration, I think we should  have 2 types or choices (mostly for felts and integers) here:
1. Do we want values in hex or decimal.
2. If it's a felt, do we want to add the string representation of it (even if we're inside an array).

This can be done later, just wanted to point out that it's concerning only cainome logic so, may be better to move it there. :+1:

glihm reviewed (commented):
Awesome, thanks for the first draft of this! Definitely aligned. :rocket:

I left a comment about a function, other than that will take a deeper look after the cleanup and testing.

But running sozo locally to test that, seems all good for now compared to the previous implementation.

Appreciate the emphasis on testing you've started. :+1:

Matth26 commented:
Thanks for the review !

> But running sozo locally to test that, seems all good for now compared to the previous implementation.

Yes I've checked on events I had locally to be the same as before, but couldn't test them all (hence the test for the `Span`)
And I fixed the custom event code which was not working if the event had more than 2 keys (which is the case in my game)

> To allow some configuration, I think we should have 2 types or choices (mostly for felts and integers)

Sure, I did it quick and dirty with the is_nested but a proper config would be better

> This can be done later, just wanted to point out that it's concerning only cainome logic so, may be better to move it there. 👍

Totally agree, I'll move this in cainome repo 👍

Matth26 commented:
I haven't added the `print_value` method to `CoreBasic` yet, as I wanted to know how to manage the `cainome` version.
However, the code is otherwise ready for review. I've cleaned up the code and added tests.

kariy commented on bin/sozo/src/commands/events.rs:
```diff
@@ -73,43 +76,49 @@ impl EventsArgs {
-                    abi::EventKind::Enum { .. } => (),
                 }
-            }
+                Err(e) => println!("Error parsing ABI: {}", e),
```
Let's have this as `anyhow::Error` and have it propagated instead.

kariy commented on bin/sozo/src/ops/events.rs:
```diff
@@ -40,124 +39,437 @@ pub async fn execute(
+                let result: Result<_, &'static str> = match inner.kind {
+                    CompositeInnerKind::Data => data.pop_front().ok_or("Missing data value"),
+                    CompositeInnerKind::Key => keys.pop_front().ok_or("Missing key value"),
+                    _ => Err("Unsupported inner kind encountered"),
```
Bcs the err is ignored later, it makes this error type pointless.

kariy commented on bin/sozo/src/ops/events.rs:
```diff
@@ -40,124 +39,437 @@ pub async fn execute(
-            ret.push_str(&format!("Model name: {}\n", name));
-        }
+        if let Token::Composite(composite) = e {
+            for inner in &composite.inners {
```
Can we move this for loop into its own function that returns error and just continue based on that function output, instead of using `continue 'outer` to manage the control flow? 

It's a bit hard to follow on first glance when reading this whole function.

kariy commented on bin/sozo/src/ops/events.rs:
```diff
@@ -40,124 +39,437 @@ pub async fn execute(
+        "u8" | "u16" | "u32" | "u64" | "u128" | "usize" | "i8" | "i16" | "i32" | "i64" | "i128" => {
+            Ok(value.to_string())
+        }
+        _ => Err(format!("Unsupported CoreBasic type: {}", cb.type_name())),
```
```suggestion
        _ => Err(anyhow!("Unsupported CoreBasic type: {}", cb.type_name())),
```
lets use the `anyhow!` macro when returning an error

kariy commented on bin/sozo/src/ops/events.rs:
```diff
@@ -40,124 +39,437 @@ pub async fn execute(
+    cb: &CoreBasic,
+    value: &FieldElement,
+    include_felt_string: bool,
+) -> Result<String, String> {
```
```suggestion
) -> Result<String> {
```

kariy reviewed (changes requested):
Good work on this @Matth26!

Left some comments after briefly went through it.

Matth26 commented:
Thanks @kariy for the review. I've updated the error handling based on your suggestions, and I think it's better now. Please let me know your thoughts 👍

glihm commented on bin/sozo/src/ops/events.rs:
```diff
@@ -40,124 +39,455 @@ pub async fn execute(
+        let event = EmittedEvent {
+            keys: vec![starknet_keccak("TestEvent".as_bytes())],
+            data: vec![
+                FieldElement::from_hex_be("0x5465737431").expect("Invalid hex"),
```
In testing you can simplify when you know that the input is expected to be correct. :+1: 
```suggestion
                FieldElement::from_hex_be("0x5465737431").unwrap(),
```

glihm commented on bin/sozo/src/ops/events.rs:
```diff
@@ -40,124 +39,455 @@ pub async fn execute(
+            let hex = format!("{:#x}", value);
+            match parse_cairo_short_string(value) {
+                Ok(parsed) if !parsed.is_empty() && include_felt_string => {
+                    Ok(format!("{} \"{}\"", parsed, hex))
+                }
+                _ => Ok(format!("\"{}\"", hex)),
```
Wdyt about having the first always being `hex` without the double quote. And if it's parseable, we output the short string in double quotes?
```suggestion
                    Ok(format!("{} \"{}\"", hex, parsed))
                }
                _ => Ok(format!("{}", hex)),
```

Matth26 replied:
Good idea !

glihm reviewed (commented):
Thanks for the work here @Matth26!
Some minor comments. I'll probably check how we can integrate the printing of the core token into cainome. Keep you posted.

glihm reviewed (approved):
Thank you for the effort @Matth26 and the time dedicated to that!
Cainome integration can be done later, and it looks great for a first rework on event support with cainome, good job! :rocket:

kariy commented on bin/sozo/src/ops/events.rs:
```diff
@@ -66,191 +66,201 @@ pub async fn execute(
+            Ok(Some(e)) => {
+                println!("{e}");
+            }
+            Ok(None) => {
+                println!("No matching event found for {:?}", event);
+            }
+            Err(e) => {
+                println!("Error parsing event: {}", e);
```
this should be returned as errors.

kariy reviewed (changes requested):
looks good to me... just need to address the one thing and then we're good

Matth26 commented:
I've just added the error as @kariy suggested

glihm reviewed (approved):
Thanks @Matth26 for the work here!
As seen with @kariy we may enhance how the ABI is retrieved from a file (currently being the scarb artifacts). But this is for a next PR.
